src/models/IContract.js

```diff
@@ -15,32 +18,29 @@ import Contract from '../utils/Contract';
  */
 class IContract {
   constructor({
-    web3,
-    contractAddress = null /* If not deployed */,
+    web3Connection = null, // Web3Connection if exists, otherwise create one from the rest of params
+    contractAddress = null, // If not deployed
     abi,
-    acc,
     tokenAddress,
+    ...params
   }) {
     try {
       if (!abi) {
         throw new Error('No ABI Interface provided');
       }
-      if (!web3) {
-        throw new Error('Please provide a valid web3 provider');
-      }
 
-      this.web3 = web3;
+      this.web3Connection = !web3Connection
+        ? new Web3Connection(params)
+        : web3Connection;
 
-      if (acc) {
-        this.acc = acc;
-      }
       this.params = {
-        web3,
+        web3Connection: this.web3Connection,
         abi,
         contractAddress,
         tokenAddress,
-        contract: new Contract(web3, abi, contractAddress),
       };
+
+      if (this.web3Connection.test) this._loadDataFromWeb3Connection();
```

Comment:
please don't add side effects to a constructor. if the user _needs_ this to happen, they will call it themselves.

Reply:
introducing "if we are testing" code just makes us off-shot that much from the main implementation when testing, I'd rather we have to mimic the implementation when testing than having shortcuts tho